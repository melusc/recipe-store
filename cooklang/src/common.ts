import {
	stringifyCookware,
	stringifyIngredient,
	stringifyTimer,
} from './schema/cooking-items.js';
import {cooklangSectionSchema, type CooklangSection} from './schema/index.js';

type FallibleResult = {
	free(): void;
	value: string;
	error: string;
};

type Parser = {
	free(this: Parser): void;
	parse(this: Parser, input: string): FallibleResult;
};

type ParserClass = {
	new (): Parser;
};

export class ParseError extends Error {}

let parser: Parser | undefined;

export function parseSection(section: string, Parser: ParserClass) {
	// Collapse all whitespace to single whitespace
	// Collapse all newlines to ["\", "\n"]
	section = section.replaceAll(/\s+/g, s => (s.includes('\n') ? '\\\n' : ' '));

	// No need for these features, escaping with backslash disables them
	// `=` is used to indicate section titles
	// `>` is used for notes
	section = section.replaceAll(/[=>]/g, String.raw`\$&`);

	parser ??= new Parser();

	const result = parser.parse(section);

	const {error, value} = result;

	if (error.trim()) {
		throw new ParseError(error);
	}

	result.free();

	const parsedResult = JSON.parse(value) as unknown;
	const errorResult = parsedResult as {error?: boolean};
	if ('error' in errorResult && errorResult['error'] === true) {
		throw new ParseError('Unknown error');
	}

	return cooklangSectionSchema.parse(parsedResult);
}

export function stringifySection(section: CooklangSection) {
	const result: string[] = [];
	const {cookware, ingredients, timers} = section;

	for (const item of section.steps) {
		switch (item.type) {
			case 'text': {
				result.push(item.value);
				break;
			}
			case 'cookware': {
				result.push(stringifyCookware(cookware[item.index]!));
				break;
			}
			case 'timer': {
				result.push(stringifyTimer(timers[item.index]!));
				break;
			}
			case 'ingredient': {
				result.push(stringifyIngredient(ingredients[item.index]!));
				break;
			}
		}
	}

	return result.join('').trim();
}
