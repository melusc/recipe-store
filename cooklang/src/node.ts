import * as common from './common.js';

import {Parser} from '#wasm/node';

export {stringifySection} from './common.js';
export type {CooklangSection} from './schema/index.js';

export function parseSection(step: string) {
	return common.parseSection(step, Parser);
}
