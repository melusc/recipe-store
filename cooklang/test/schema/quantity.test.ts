import {test, assert, expect} from 'vitest';

import {ParseError} from '../../src/common.js';
import {parseSection} from '../../src/node.js';
import {
	stringifyQuantity,
	quantitySchema,
	type Quantity,
	unitlessQuantitySchema,
	type UnitlessQuantity,
} from '../../src/schema/quantity.js';

test.for([
	['one%slice', 'one slice'],
	['one slice', 'one slice'],
	['100*', '100'],
	['100*%g', '100 g'],
	['100', '100'],
	['100%m', '100 m'],
	['1.5%ft', '1.5 ft'],
	['1/2', '1/2'],
	['1/2%L', '1/2 L'],
	['1 1/2', '1 1/2'],
	['1 1/2%L', '1 1/2 L'],
	['1 300/401', '1 300/401'],
	['1 300/401%L', '1 300/401 L'],
	['1-2', '1-2'],
	['1-2%tbsp', '1-2 tbsp'],
])('Ingredient quantity stringify("%s") -> "%s"', ([input, output]) => {
	const cooklangSource = `@ingredient{${input}}`;
	const parsed = parseSection(cooklangSource) as {
		ingredients: Array<{quantity: Quantity}>;
	};
	assert('ingredients' in parsed);
	assert.containsAllKeys(parsed, ['ingredients']);
	assert.lengthOf(parsed.ingredients, 1);
	let quantity!: Quantity;
	assert.doesNotThrow(() => {
		quantity = quantitySchema.parse(parsed.ingredients[0]!.quantity);
	});

	assert.equal(stringifyQuantity(quantity), output);
});

test.for([
	['100%min', '100 min'],
	['1.5%h', '1.5 h'],
	['1/2%h', '1/2 h'],
	['1 1/2%s', '1 1/2 s'],
	['1 300/401%d', '1 300/401 d'],
])('Timer quantity stringify("%s") -> "%s"', ([input, output]) => {
	const cooklangSource = `~{${input}}`;
	const parsed = parseSection(cooklangSource) as {
		timers: Array<{quantity: Quantity}>;
	};
	assert('timers' in parsed);
	assert.containsAllKeys(parsed, ['timers']);
	assert.lengthOf(parsed.timers, 1);
	let quantity!: Quantity;
	assert.doesNotThrow(() => {
		quantity = quantitySchema.parse(parsed.timers[0]!.quantity);
	});

	assert.equal(stringifyQuantity(quantity), output);
});

test.for([
	// Unitless
	'100',
	// Not time unit
	'100%m',
	// Not number
	'ten%minute',
	'1-2%minute',
])('Invalid timer quantity %j', timerInput => {
	const cooklangSource = `~{${timerInput}}`;
	assert.throws(() => {
		parseSection(cooklangSource);
	}, ParseError);
});

test.for([
	['one', 'one'],
	['100', '100'],
	['1/2', '1/2'],
	['1 1/2', '1 1/2'],
	['1 300/401', '1 300/401'],
	['1-2', '1-2'],
])('Cookware quantity stringify("%s") -> "%s"', ([input, output]) => {
	const cooklangSource = `#a{${input}}`;
	const parsed = parseSection(cooklangSource) as {
		cookware: Array<{quantity: UnitlessQuantity}>;
	};
	assert('cookware' in parsed);
	assert.containsAllKeys(parsed, ['cookware']);
	assert.lengthOf(parsed.cookware, 1);
	expect(parsed.cookware[0]).toMatchSnapshot();
	let quantity!: UnitlessQuantity;
	assert.doesNotThrow(() => {
		quantity = unitlessQuantitySchema.parse(parsed.cookware[0]!.quantity);
	});

	assert.equal(stringifyQuantity(quantity), output);
});

test.for(['1%g', 'one%g'])('Invalid cookware quantity %j', cookwareInput => {
	const cooklangSource = `#a{${cookwareInput}}`;
	assert.throws(() => {
		parseSection(cooklangSource);
	}, ParseError);
});
