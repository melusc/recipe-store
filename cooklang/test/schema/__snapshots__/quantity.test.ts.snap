// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Cookware quantity stringify("1 1/2") -> "1 1/2" 1`] = `
{
  "alias": null,
  "name": "a",
  "note": null,
  "quantity": {
    "type": "fixed",
    "value": {
      "type": "number",
      "value": {
        "type": "fraction",
        "value": {
          "den": 2,
          "err": 0,
          "num": 1,
          "whole": 1,
        },
      },
    },
  },
}
`;

exports[`Cookware quantity stringify("1 300/401") -> "1 300/401" 1`] = `
{
  "alias": null,
  "name": "a",
  "note": null,
  "quantity": {
    "type": "fixed",
    "value": {
      "type": "number",
      "value": {
        "type": "fraction",
        "value": {
          "den": 401,
          "err": 0,
          "num": 300,
          "whole": 1,
        },
      },
    },
  },
}
`;

exports[`Cookware quantity stringify("1/2") -> "1/2" 1`] = `
{
  "alias": null,
  "name": "a",
  "note": null,
  "quantity": {
    "type": "fixed",
    "value": {
      "type": "number",
      "value": {
        "type": "fraction",
        "value": {
          "den": 2,
          "err": 0,
          "num": 1,
          "whole": 0,
        },
      },
    },
  },
}
`;

exports[`Cookware quantity stringify("1-2") -> "1-2" 1`] = `
{
  "alias": null,
  "name": "a",
  "note": null,
  "quantity": {
    "type": "fixed",
    "value": {
      "type": "text",
      "value": "1-2",
    },
  },
}
`;

exports[`Cookware quantity stringify("100") -> "100" 1`] = `
{
  "alias": null,
  "name": "a",
  "note": null,
  "quantity": {
    "type": "fixed",
    "value": {
      "type": "number",
      "value": {
        "type": "regular",
        "value": 100,
      },
    },
  },
}
`;

exports[`Cookware quantity stringify("one") -> "one" 1`] = `
{
  "alias": null,
  "name": "a",
  "note": null,
  "quantity": {
    "type": "fixed",
    "value": {
      "type": "text",
      "value": "one",
    },
  },
}
`;
